feat: Add Hackathon article to blog for Matrix Bridge (#523)
wazo-platform/wazo-platform.org: content/blog/hackathon-2022-wazo-matrix-bridge.md

## Possible improvements

A side from simply completing the other side of the bridge there are many other things we wanted to add. We also wanted to handle files and other media content as well as reactions. The Mautrix library also has tools for using chatbots to perform the initial configuration and authentication of bridged users, which would have been ideal. In our case, it was simply configured in the config file.
Nonetheless, it was a very interesting project.

## What we learned

We learned a lot about how Mautrix works and, more generally, the things we could do with a bridge. All platforms are a bit different and provide different tools for linking them to other services, but in a lot of cases it is possible to do something like this to allow for different platforms to be linked together. Its always nice when you can make platforms work together.

We also got a lot of ideas of things that it would be nice to add to improve chatd, maybe sometime in the future.

> The code for the bridge is [located here](https://github.com/sopelj/mautrix-wazo), but keep in mind it was juste a test for the Hackathon and is not fully functional.

Reviewer: Juste -> just

Author: Oops, thanks!

Title: Hackathon 2022 - Bridging Wazo and Matrix
Date: 2022-10-20 12:00:00
Author: Jesse Sopel
Category: Hackathon
Tags: Chat, Python, Matrix, Bridge, Hackathon
Slug: hackathon-2022-bridging-wazo-and-matrix
Status: published

> # Hackathon 2022 - Bridging Wazo and Matrix

Reviewer 2: No need to add the main title in the content, the website will that the title in meta-data (aka line #1)

Author: Oh thanks. I wasn't sure how it was rendered.

Reviewer 2: FYI: I did the change since I reformated the header with the new syntax.

Author: Oh perfect, thanks!

Every year Wazo organizes a Hackathon with their employees, and as many of us as possible get together to work on some fun projects and see what we can learn.

> There were multiple teams working on different projects. Our team consisted of Charles, Francis, Francois and I (Jesse).

Reviewer 3: maybe you could put the links to their github profiles :)

Author: Good idea. I'll try and find them.

Author: Added!

> We decided to both work on creating a bridge between Wazo's built-in chat system and a Matrix server, as well as [testing out adding some new features to the existing chat](https://wazo-platform.org/blog/hackaton-2022-building-group-chat-with-wazo), but this article is focused on the former. 

Reviewer 2: To keep in mind, if the other article is not yet ready, maybe the link should added in another PR

Author: Good point. I assumed they'd be released at the same time, but I'm in no rush to publish it. We can wait until yours is published if you want too.

## The Objective

The goal was to create a bidirectional bridge between Wazo's [Chatd](https://wazo-platform.org/documentation/overview/chat.html) and a [Matrix](https://matrix.org/) Homeserver.

> A user that exists on both platforms can have the two accounts mapped to each other, but in case they do not exist or have no account configured the bridge then "puppet" accounts should be created for them. This allows the Matrix user to chat with any user on Wazo as if they were using it directly. This is known as a [Double Puppeting](https://matrix.org/docs/guides/types-of-bridging#double-puppeted-bridge). Now, we didn't quite get there, but we made some good progress and learned some new things.

Reviewer 4: ```suggestion
A user that exists on both platforms can have the two accounts mapped to each other, but in case they do not exist or have no account configured on the bridge, then "puppet" accounts should be created for them. This allows the Matrix user to chat with any user on Wazo as if they were using it directly. This is known as a [Double Puppeting](https://matrix.org/docs/guides/types-of-bridging#double-puppeted-bridge). Now, we didn't quite get there, but we made some good progress and learned some new things.
```

## Technologies/Libraries used

- [Chatd](https://wazo-platform.org/documentation/overview/chat.html)- Chat messages
- [Webhookd](https://wazo-platform.org/documentation/overview/webhook.html) - Listen for messages and forward to bridge
- [Confd](https://wazo-platform.org/documentation/overview/configuration.html) - Get user information
- [Matrix Synapse](https://matrix.org/docs/projects/server/synapse) - The Matrix homeserver
- [Mautrix](https://github.com/mautrix/python) - Library for the bridge

> - [aiohttp](https://docs.aiohttp.org/en/stable/)  - API calls and server for webooks

Reviewer 4: ```suggestion
- [aiohttp](https://docs.aiohttp.org/en/stable/)  - API calls and server for webhooks
```

## Step 1: Understanding, planning and the database

Our first step, but to figure out how Mautrix works. Their bridges are very feature rich, but we were unable to find much documentation for the actual library, so we needed to spend a lot of time reading the library code and looking at other bridges to figure out how they worked. Also, lots of testing was needed. Then we needed to make the Database models for the different items that need to be saved in the bridge, such as Wazo-Matrix mappings for users and groups. Typically, we use PostgreSQL, but for our test we simply used sqlite because it required less configuration and it would be easier to copy the database around for our tests.

## Step 2: Bridge Wazo to Matrix

> In order to test something, we needed at least one direction to work. So we started with the Wazo side. We needed to know as soon as a message was sent to our user so we could create the groups/users and forward the message to Matrix. To do this, we created a small HTTP server using aiohttp and set up a webhook with Webhookd. The webhook listens for an new message event in chatd and then posts the message information to a URL that our bridge exposes. When the message content is received the bridge checks to whom the message is being sent. If the users don't already in the bridge and Matrix, then it must create them and the room. The user of Matrix receives and invite to the new room which contains puppeted users. Once the room and users already exist, the bridge simply forwards the new messages to the correct group in Matrix.

Reviewer 4: ```suggestion
In order to test something, we needed at least one direction to work. So we started with the Wazo side. We needed to know as soon as a message was sent to our user so we could create the groups/users and forward the message to Matrix. To do this, we created a small HTTP server using aiohttp and set up a webhook with Webhookd. The webhook listens for a new message event in chatd and then posts the message information to a URL that our bridge exposes. When the message content is received the bridge checks to whom the message is being sent. If the users are not already in the bridge and Matrix, then it must create them and the room. The user of Matrix receives an invite to the new room which contains puppeted users. Once the room and users already exist, the bridge simply forwards the new messages to the correct group in Matrix.
```

## Step 3: Different message formats

> Parallel to the bridge, we were also working on supporting new types of content in chatd such as Markdown, emoji, files and reactions. ([You can read about that here](https://wazo-platform.org/blog/hackaton-2022-building-group-chat-with-wazo)). The Matrix protocol already supports these things as does the client we were using (Element). In order to support Markdown, we simply used a python library python-markdown to render the text into HTML and send that as the `formatted_body` to Matrix. This allowed for rich text content such as lists, tables, bold, italic, strike-through, embedded images and more. As for emoji, since they are simply unicode characters, no special logic was needed. However, sadly, we did not have time to add reactions or file handling to the bridge during the event.

Reviewer 4: ```suggestion
Parallel to the bridge, we were also working on supporting new types of content in chatd such as Markdown, emoji, files and reactions. ([You can read about that here](https://wazo-platform.org/blog/hackaton-2022-building-group-chat-with-wazo)). The Matrix protocol already supports these things as does the client we were using (Element). In order to support Markdown, we simply used a python library python-markdown to render the text into HTML and send that as the `formatted_body` to Matrix. This allowed for rich text content such as lists, tables, bold, italic, strike-through, embedded images and more. As for emojis, since they are simply Unicode characters, no special logic was needed. However, sadly, we did not have time to add reactions or file handling to the bridge during the event.
```

Author: Wow, I'm embarrassed you found so many typos. Thanks! I merged all the others, but I'm pretty sure Unicode is supposed to be capitalised since it is a proper noun and emoji is a Japanese loan word and as such shouldn't really be pluralised either.

> Parallel to the bridge, we were also working on supporting new types of content in chatd such as Markdown, emoji, files and reactions. ([You can read about that here](https://wazo-platform.org/blog/hackaton-2022-building-group-chat-with-wazo)). The Matrix protocol already supports these things as does the client we were using (Element). In order to support Markdown, we simply used a python library python-markdown to render the text into HTML and send that as the `formatted_body` to Matrix. This allowed for rich text content such as lists, tables, bold, italic, strike-through, embedded images and more. As for emoji, since they are simply unicode characters, no special logic was needed. However, sadly, we did not have time to add reactions or file handling to the bridge during the event.

Reviewer 2: (ditto/reminder) about second article if not ready before merge

## Step 4: Bridge Matrix to Wazo

The next step would have been to send the messages written in Matrix to chatd via the API. We got fairly far along on this, but sadly did not complete it in time, so it was not functional for our demo and the bridge remained one-sided. So any messages written in chatd were correctly received in the Matrix client, but it was not possible to respond in the client.

## Possible improvements

> A side from simply completing the other side of the bridge there are many other things we wanted to add. We also wanted to handle files and other media content as well as reactions. The Mautrix library also has tools for using chatbots to perform the initial configuration and authentication of bridged users, which would have been ideal. In our case, it was simply configured in the config file.

Reviewer 4: ```suggestion
Aside from simply completing the other side of the bridge there are many other things we wanted to add. We also wanted to handle files and other media content as well as reactions. The Mautrix library also has tools for using chatbots to perform the initial configuration and authentication of bridged users, which would have been ideal. In our case, it was simply configured in the config file.
```

Nonetheless, it was a very interesting project.

## What we learned

We learned a lot about how Mautrix works and, more generally, the things we could do with a bridge. All platforms are a bit different and provide different tools for linking them to other services, but in a lot of cases it is possible to do something like this to allow for different platforms to be linked together. It's always nice when you can make platforms work together.

> We also got a lot of ideas of things that it would be nice to add to improve chatd, maybe sometime in the future.

Reviewer 5: ```suggestion
We also got a lot of ideas of things that would be nice to add or improve in chatd, maybe sometime in the future.
```
